pkg/planner/core/plan_cache.go

```diff
@@ -585,7 +660,62 @@ func rebuildRange(p Plan) error {
 				}
 			}
 		}
+		// TODO: fix TableDual!
+		if len(x.PartitionDefs) > 0 {
```

Comment:
Can we put the operation of updating PointGet.PartitionDef and BatchPointGet.PartitionDef to the executor builder?  Just like all other Readers, then they have the same process to handle partitioning tables: 1) Re-Eval some expressions at the optimization phase, 2) Re-Partition Prunning at the executor builder phase.

Reply:
I have made a major code move/refactor to only re-evaluate the ParamMarkers in rebuildRange, and move the (re)partition pruning to the executor building phase.

By moving the partition pruning to the Executor building phase, several changes was needed:
- EXPLAIN now needs to build the executor, even if not EXPLAIN ANALYZE, to get the partition pruning correct.
- We can remove a lot of code for partition pruning for `[Batch]PointGet`, which only covered a smaller set of partitioning, and replace it with the more commonly used `GetPartition[Idx]ByRow()`/`LocatePartition()` functionality (which is used when you have the actual values to check which partition a row belongs to, and is used in the write path).

And some side effects happened:
- EXPLAIN now starts a transaction (to get the current schema version?)
- the [Batch]PointGet plan will be explained as PointGet with `partition:dual` instead of TableDual plan (the executor is still TableDual, but EXPLAIN uses the plan, not the executor)

I'm still updating test results for various test to reflect these changes, but before I'm done, I just wanted to check with you, @qw4990, if you think this is the right way?